packages/core/src/indexing/index-api.ts

```diff
@@ -4,14 +4,47 @@ import type {
   Page,
   Pagination,
   StreamState,
+  DiagnosticsLogger,
 } from '@ceramicnetwork/common'
+import { SyncOptions } from '@ceramicnetwork/common'
 import type { DatabaseIndexApi } from './database-index-api.js'
-import { NotImplementedError } from './not-implemented-error.js'
+import type { Repository } from '../state-management/repository.js'
 
+/**
+ * API to query an index.
+ */
 export class IndexApi implements IndexClientApi {
-  constructor(private readonly databaseIndexApi: DatabaseIndexApi | undefined) {}
+  constructor(
+    private readonly databaseIndexApi: DatabaseIndexApi | undefined,
+    private readonly repository: Repository,
+    private readonly logger: DiagnosticsLogger
+  ) {}
 
-  queryIndex(query: BaseQuery & Pagination): Promise<Page<StreamState>> {
-    throw new NotImplementedError('IndexApi::queryIndex')
+  /**
+   * Query the index. Ask an indexing database for a list of StreamIDs,
+   * and convert them to corresponding StreamState instances via `Repository::streamState`.
+   *
+   * We assume that a state store always contains StreamState for an indexed stream.
+   */
+  async queryIndex(query: BaseQuery & Pagination): Promise<Page<StreamState>> {
+    if (this.databaseIndexApi) {
+      const page = await this.databaseIndexApi.page(query)
+      const streamStates = await Promise.all(
```

Comment:
I'm a little worried about us hammering S3 with hundreds of simultaneous requests.  I think we might want to break this down into smaller batches and/or rate limit how fast we hammer the state store with these requests

Reply:
I believe it is a responsibility of S3StateStore then, and not this thing. Created an issue for that.